### PR TITLE
Support sequences + most types used in spark

### PR DESCRIPTION
**Pull request by michael72**
Hello @vincenzobaz

I wanted to evaluate use of Scala 3 in our project that uses spark and came across your project and found it very useful.
However code generation did not work entirely, since we also rely on collections in our case classes used in spark.
Hence I've tried my best to add most of the datatypes that are supported in spark sql.

Maybe you can have a look at the PR - sorry it is quite big. I also added sbt-scalafmt which makes it possible to use scalafmtAll and scalafmtSbt inside sbt.

There is one long-ish and quirky matching expression at the end of Serializer and Deserializer in summonAll which makes it possible now to encode and decoder very long data types. I know that it is generally possible to do so when increasing "-Xmax-inlines:256" for example - but with this "bulk-processing" of the tuples that I added: I don't need that setting in our project and it doesn't crash the compiler with a stackoverflow. I tried to fix that stackoverflow but it seems the best way to do so is to restrict the stack size of that summonAll expression otherwise the compiler will crash occasionally in our project, especially when clean building.

I also added myself as a developer - I hope that is OK.

I also hope that this PR is OK - or give me your thoughts and please review. Thanks!

Regards
Michael

btw - this is the 2nd try for the PR - this time with the correct branches (I hope)

### CODE REVIEW

File: examples/src/main/scala/sql/StarWars.scala
Comment (michael72) [inline examples/src/main/scala/sql/StarWars.scala:62]: I don't think this was working before

File: build.sbt
Comment (vincenzobaz) [inline build.sbt:91]: Do you wish to become a maintainer or more of a regular contributor?

**Review by vincenzobaz — COMMENTED** [review]
Very cool work! I am very thankful for the tests too.

File: build.sbt
Comment (michael72) [reply build.sbt:91]: not sure - it is your call. :smile_cat:

File: encoders/src/main/scala/scala3encoders/derivation/Deserializer.scala
Comment (michael72) [reply encoders/src/main/scala/scala3encoders/derivation/Deserializer.scala:247]: Hm - 16 was just a number - if you look below - the number of maximum inlines was (originally) 32 - so 8 might have been a good number as well - or 32. 8 might actually improve the readability - and should (maybe) also be enough - at least for our project it is enough.

so - in our _big_ project I regularly got stackoverflows of the compiler when compiling a certain sub project. It turned out the reason was some bigger case class with > 256 members including sub-case-classes (by joining tables).
Before having this stackoverflow however I had that error message in the spark-scala3 project - if you just comment out those lines above you will also get with `sbt clean test`:

```log
[error] -- Error: /home/michael/projects/spark-scala3/encoders/src/test/scala/sql/EncoderDerivationSpec.scala:346:38 
[error] 346 |    val encoder = summon[Encoder[Big]]
[error]     |                                      ^
[error]     |                   Maximal number of successive inlines (32) exceeded,
[error]     |                   Maybe this is caused by a recursive inline method?
[error]     |                   You can use -Xmax-inlines to change the limit.
[error]     |---------------------------------------------------------------------------
[error]     |Inline stack trace
[error]     |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
[error]     |This location contains code that was inlined from EncoderDerivationSpec.scala:346
```

however when adding this -Xmax-inlines:256 (256 was just a guess) to our project - then we would get stackoverflows by the compiler which weren't really helpful.

So: to improve readability I would reduce to 8 - but on the other hand that increases the risk of getting stackoverflows when summoning the encoder for large case classes (which might just be a representation for joined tables).

File: encoders/src/main/scala/scala3encoders/derivation/Deserializer.scala
Comment (michael72) [reply encoders/src/main/scala/scala3encoders/derivation/Deserializer.scala:166]: OK - I just took over your other branch which was supposed to add support for Maps etc. - and that comment was probably left - but you're right - that would be a tad overkill :-)

File: build.sbt
Comment (michael72) [reply build.sbt:19]: yes - as stated in the comment
if you look at
 https://github.com/apache/spark/blob/v3.3.2/launcher/src/main/java/org/apache/spark/launcher/JavaModuleOptions.java

```
required by Spark when using Java 17. `DEFAULT_MODULE_OPTIONS` has added
 `-XX:+IgnoreUnrecognizedVMOptions` to be compatible with Java 8 and Java 11.
```

File: encoders/src/main/scala/scala3encoders/derivation/Deserializer.scala
Comment (michael72) [reply encoders/src/main/scala/scala3encoders/derivation/Deserializer.scala:166]: but I can certainly help as well if I'm a maintainer

File: build.sbt
Comment (michael72) [reply build.sbt:91]: but I can certainly help as a maintainer

File: encoders/src/main/scala/scala3encoders/derivation/Deserializer.scala
Comment (vincenzobaz) [reply encoders/src/main/scala/scala3encoders/derivation/Deserializer.scala:247]: Thanks for explaining :)

**Review by vincenzobaz — APPROVED** [review]

**Comment by michael72** [comment]
Hey @vincenzobaz,

Sooo.... are we gonna merge this? :-)

I could maybe try to work on Enums or on udf (or replacement of udf-call) next (in time). That could be tricky but it could probably be done.

As discussed, if you like, you could add me as a maintainer

**Comment by michael72** [comment]
fixes #14

**Comment by michael72** [comment]
> I would say udf have higher priority, but I let you focus on what is more fun for you!

that's nice - unfortunately I do not have the rights to do so - I can only close the PR but not merge it

**Comment by vincenzobaz** [comment]
> > I let you do the honor of merging!
> 
> @vincenzobaz that's nice - unfortunately I do not have the rights to do so - I can only close the PR but not merge it

I am sorry, I thought the approval is enough. I will look into how to adding a contributor to the repo :smile: . I am not that familiar with this part of github

**Pull request merged**
